### PR TITLE
Remove `pyproject.toml`

### LINKED ISSUE #4727: `pip install` completely ignores existing source builds and installed dependencies
I suspect this has to do with #4619. 

I am on the latest master, and now every time I call `pip install -v -e .` two things happens:
1. These packages keeps being reinstalled despite I already have them in my env: setuptools, wheel, Cython, fastrlock
2. All modules are re-cythonized and recompiled from scratch, despite they've been built and nothing has changed 

I will try to build in a fresh env to see if something is wrong with my current env. But it's better to be confirmed independently.

cc: @kmaehashi @emcastillo

### PR DESCRIPTION
**Pull request by kmaehashi**
Closes #4727

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -24,8 +24,7 @@
 
 
 requirements = {
-    # setup_requires remains here for pip v18 or earlier.
-    # Keep in sync with pyproject.yaml.
+    # TODO(kmaehashi): migrate to pyproject.yaml (see #4727, #4619)
```
Comment (toslunar) [inline setup.py:27]:
```suggestion
    # TODO(kmaehashi): migrate to pyproject.toml (see #4727, #4619)
```

Comment (kmaehashi) [reply setup.py]: Nice catch!

**Review by toslunar — APPROVED** [review]

**Comment by toslunar** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Thanks for handling my last minute request @kmaehashi 🙏

**Comment by leofang** [comment]
FFT callback tests failed?! Is it new?

**Comment by kmaehashi** [comment]
`cupy-py36` now changed to use CUDA 11.2 https://github.com/chainer/chainer-test/pull/635 so that can be a reason...

**Comment by leofang** [comment]
Hmmm I thought I checked with 11.2. Let me recheck.

**Comment by leofang** [comment]
Tests passed fine locally, but I'm in a dev env. Based on the error log it looks like some Cython files are missing so cythonize fails. I saw other CI runs also failed with callbacks, so it's probably not due to removing `pyproject.toml`.

**Comment by kmaehashi** [comment]
The reason for these failures is coming from the design of cufft callback. CuPy installation does not contain pxd files although `cufft.pxd` is `cimport`ing `memory.pxd`, that's why Cython errors are raised.
Previously, chainer-test was testing against CuPy installed via `pip install -e` , so all pxd files were visible. However, https://github.com/chainer/chainer-test/pull/634 changed tests to runs against CuPy installed to site-packages, so pxd files are now invisible from the test script.

CUDA 11.2 is not related to these failure. For CUDA 11.1 or earlier, no cythonize is running on Jenkins because `callback_cache` is already generated.

I think it's better to skip all cufft callback tests for this week's release, and discuss what we can do to support cufft callback for the March release.

**Comment by leofang** [comment]
Thanks, that makes sense, though I was under the impression that we included *.pxd/*.pyx in both sdist and wheel, so shouldn't it just work? I could be wrong though.

**Comment by kmaehashi** [comment]
> we included *.pxd/*.pyx in both sdist and wheel, so shouldn't it just work?

Technically, yes. But once we do that people will start using Cython API from external libraries (e.g. #79, #130), in that case, we have to be very very careful to avoid breaking the compatibility because Cython API breakage will cause the library import to fail immediately. Maintaining Cython API is much costly compared to Python API. That's why we don't expose Cython level API since the initial release.

**Comment by leofang** [comment]
> Maintaining Cython API is much costly compared to Python API. That's why we don't expose Cython level API since the initial release.

I agree. What if we explicitly state in the docs etc that the Cython API is meant to be private and we do not guarantee stability across even minor releases? I think it'd also help address general questions.

I am testing this locally. I think it's easiest if we could include the following `.pxd` files:
```
    'cupy_backends/cuda/api/__init__.pxd',
    'cupy_backends/cuda/api/driver.pxd',
    'cupy_backends/cuda/api/runtime.pxd',
    'cupy/cuda/__init__.pxd',
    'cupy/cuda/memory.pxd',
    'cupy/cuda/stream.pxd',
```

**Comment by kmaehashi** [comment]
@toslunar Could you proceed to merge this PR? The test failures are irrelevant.

**Pull request merged**